### PR TITLE
MULE-17799: Error when using 100-Continue and Proxy

### PR DESCRIPTION
**Pull request by eze210**

### CODE REVIEW

File: modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java [hunk 1]
```diff
@@ -114,17 +115,59 @@ public void testHeadersN() throws Exception {
...
-    
+
     public void testDecoder100continueThen200() {
         try {
-            doTestDecoder("HTTP/1.1 100 Continue\n\nHTTP/1.1 200 OK\n\n", 4096);
-            assertTrue(true);
+            HttpPacket packet = doTestDecoder("HTTP/1.1 100 Continue\n\nHTTP/1.1 200 OK\n\n", 4096);
+            assertTrue(packet.getHttpHeader() instanceof HttpResponsePacket);
+            HttpResponsePacket response = (HttpResponsePacket) packet.getHttpHeader();
+            assertTrue(response.getStatus() == 200);
```
Comment (fsgonz) [inline modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java]: use assertEquals

File: modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java [hunk 2]
```diff
@@ -114,17 +115,59 @@ public void testHeadersN() throws Exception {
...
             assertTrue("Unexpected exception", false);
         }
     }
 
+    public void testDecoder100continueWithAHeaderThen200() {
+        try {
+            HttpPacket packet = doTestDecoder("HTTP/1.1 100 Continue\r\nConnection: keep-alive\r\n\r\nHTTP/1.1 200 OK\r\n\r\n", 4096);
+            assertTrue(packet.getHttpHeader() instanceof HttpResponsePacket);
+            HttpResponsePacket response = (HttpResponsePacket) packet.getHttpHeader();
+            assertTrue(response.getStatus() == 200);
```
Comment (fsgonz) [inline modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java]: use assertEquals

File: modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java [hunk 3]
```diff
@@ -114,17 +115,59 @@ public void testHeadersN() throws Exception {
...
+    }
+
+    public void testDecoder100continueAndResponseUsingSameConnection() throws IOException {
+        HttpClientFilter filter = new HttpClientFilter(4096);
+
+        FilterChainContext ctx = FilterChainContext.create(new StandaloneConnection());
+
+        handleRead(filter, ctx, "HTTP/1.1 100 Continue\n\n");
+        HttpPacket lastPacket = (HttpPacket) handleRead(filter, ctx, "HTTP/1.1 200 OK\n\n");
+        assertTrue(((HttpResponsePacket) lastPacket.getHttpHeader()).getStatus() == 200);
```
Comment (fsgonz) [inline modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java]: assertEquals

File: modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java [hunk 4]
```diff
@@ -114,17 +115,59 @@ public void testHeadersN() throws Exception {
...
+    }
+
+    public void testDecoder100continueAndResponseUsingSameConnectionWithHeader() throws IOException {
+        HttpClientFilter filter = new HttpClientFilter(4096);
+
+        FilterChainContext ctx = FilterChainContext.create(new StandaloneConnection());
+
+        handleRead(filter, ctx, "HTTP/1.1 100 Continue\nConnection: keep-alive\n\n");
+        HttpPacket lastPacket = (HttpPacket) handleRead(filter, ctx, "HTTP/1.1 200 OK\n\n");
+        assertTrue(((HttpResponsePacket) lastPacket.getHttpHeader()).getStatus() == 200);
```
Comment (fsgonz) [inline modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java]: assertEquals

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java [hunk 5]
```diff
@@ -114,17 +115,59 @@ public void testHeadersN() throws Exception {
...
         }
     }
 
+    public void testDecoder100continueWithAHeaderThen200() {
+        try {
+            HttpPacket packet = doTestDecoder("HTTP/1.1 100 Continue\r\nConnection: keep-alive\r\n\r\nHTTP/1.1 200 OK\r\n\r\n", 4096);
+            assertTrue(packet.getHttpHeader() instanceof HttpResponsePacket);
+            HttpResponsePacket response = (HttpResponsePacket) packet.getHttpHeader();
+            assertTrue(response.getStatus() == 200);
+        } catch (IllegalStateException e) {
```
Comment (fsgonz) [inline modules/http/src/test/java/org/glassfish/grizzly/http/HttpResponseParseTest.java]: I would remove this catch-block, add the throws for the method and let the test fail with the unexpected exception. I would change that in the previous test that uses this too.

**Review by fsgonz — APPROVED** [review]

**Review by afelisatti — APPROVED** [review]
Please validate this against the connector to make sure the 100 continue tests there are ok since we don't have good coverage on the service side.

**Pull request merged**